src/System.Linq/tests/ReverseTests.cs

```diff
@@ -14,86 +14,59 @@ public void InvalidArguments()
-        [InlineData(new int[] { 5 })]
-        [InlineData(new int[] { 1, 3, 5 })]
-        [InlineData(new int[] { 2, 4, 6, 8 })]
-        public void ReverseMatches(int[] input)
+        [MemberData(nameof(ReverseData))]
+        public void Reverse<T>(IEnumerable<T> source, T dummy)
         {
-            int[] expectedResults = new int[input.Length];
-            for (int i = 0; i < input.Length; i++)
-            {
-                expectedResults[i] = input[input.Length - 1 - i];
-            }
-
-            Assert.NotSame(input, Enumerable.Reverse(input));
-
-            Assert.Equal(expectedResults, input.Reverse());
-            Assert.Equal(expectedResults, new TestCollection<int>(input).Reverse());
-            Assert.Equal(expectedResults, new TestEnumerable<int>(input).Reverse());
-            Assert.Equal(expectedResults, new TestReadOnlyCollection<int>(input).Reverse());
+            T[] expected = source.ToArray();
+            Array.Reverse(expected);
 
-            Assert.Equal(expectedResults.Select(i => i * 2), input.Select(i => i * 2).Reverse());
-            Assert.Equal(expectedResults.Where(i => true).Select(i => i * 2), input.Where(i => true).Select(i => i * 2).Reverse());
-            Assert.Equal(expectedResults.Where(i => false).Select(i => i * 2), input.Where(i => false).Select(i => i * 2).Reverse());
-        }
-
-        [Fact]
-        public void SameResultsRepeatCallsIntQuery()
-        {
-            var q = from x in new[] { 9999, 0, 888, -1, 66, -777, 1, 2, -12345 }
-                    where x > Int32.MinValue
-                    select x;
+            IEnumerable<T> actual = source.Reverse();
 
-            Assert.Equal(q.Reverse(), q.Reverse());
-        }
-
-        [Fact]
-        public void SameResultsRepeatCallsStringQuery()
-        {
-            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
-                    where !String.IsNullOrEmpty(x)
-                    select x;
-
-            Assert.Equal(q.Reverse(), q.Reverse());
-        }
-
-        [Fact]
-        public void SomeRepeatedElements()
-        {
-            int?[] source = new int?[] { -10, 0, 5, null, 0, 9, 100, null, 9 };
-            int?[] expected = new int?[] { 9, null, 100, 9, 0, null, 5, 0, -10 };
+            Assert.Equal(expected, actual);
+            Assert.Equal(expected.Count(), actual.Count()); // Count may be optimized.
+            Assert.Equal(expected, actual.ToArray());
+            Assert.Equal(expected, actual.ToList());
 
-            Assert.Equal(expected, source.Reverse());
-        }
+            Assert.Equal(expected.FirstOrDefault(), actual.FirstOrDefault());
+            Assert.Equal(expected.LastOrDefault(), actual.LastOrDefault());
 
-        [Fact]
-        public void ToArray()
-        {
-            int?[] source = new int?[] { -10, 0, 5, null, 0, 9, 100, null, 9 };
-            int?[] expected = new int?[] { 9, null, 100, 9, 0, null, 5, 0, -10 };
```

Comment:
Nevermind, actually: They can't be fed into the `[Theory]` because xUnit cannot do type inference for nullables. (When an `int?` is boxed it's an `int` and xUnit will try to interpret the `int?[]` as an `IEnumerable<int>`.) I don't think there's a special case for nullables within `Reverse`.